src/zcash/JoinSplit.cpp

```diff
@@ -0,0 +1,344 @@
+#include "JoinSplit.hpp"
+#include "prf.h"
+#include "sodium.h"
+
+#include "zerocash/utils/util.h"
+
+#include <memory>
+
+#include <boost/foreach.hpp>
+#include <boost/format.hpp>
+#include <boost/optional.hpp>
+#include <fstream>
+#include "libsnark/common/default_types/r1cs_ppzksnark_pp.hpp"
+#include "libsnark/zk_proof_systems/ppzksnark/r1cs_ppzksnark/r1cs_ppzksnark.hpp"
+#include "libsnark/gadgetlib1/gadgets/hashes/sha256/sha256_gadget.hpp"
+#include "libsnark/gadgetlib1/gadgets/merkle_tree/merkle_tree_check_read_gadget.hpp"
+
+#include "sync.h"
+
+using namespace libsnark;
+
+namespace libzcash {
+
+#include "zcash/circuit/gadget.tcc"
+
+CCriticalSection cs_ParamsIO;
+CCriticalSection cs_InitializeParams;
+
+template<typename T>
+void saveToFile(std::string path, T& obj) {
+    LOCK(cs_ParamsIO);
+
+    std::stringstream ss;
+    ss << obj;
+    std::ofstream fh;
+    fh.open(path, std::ios::binary);
```

Comment:
Not too critical I think; we only use this code for the `GenerateParams` binary right now.